File: packages/fxa-auth-server/loadtest/loadtests.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

const assert = require('assert');
const crypto = require('crypto');
const Client = require('../client/');
const srp = require('srp');
const P = require('p-promise')


const SERVER_URL = "http://idp.loadtest.lcip.org";


// Error constants used by the picl-idp API.
//
ERROR_ACCOUNT_EXISTS = 101;
ERROR_UNKNOWN_ACCOUNT = 102;


// The tests need a public key for the server to sign, but we don't actually
// do anything with it.  It suffices to use a fixed dummy key throughout.
//
var DUMMY_PUBLIC_KEY = {
  "algorithm": "RS",
  "n": "4759385967235610503571494339196749614544606692567785790953934768202714280652973091341316862993582789079872007974809511698859885077002492642203267408776123",
  "e": "65537"
};


// We don't want to do any key-stretching during the loadtest, because it
// takes a long time.  Instead we start from a fixed set of SRP credentials
// and mix in just enough user-account-specific information to make the
// authentication work properly.
//
var DUMMY_CLIENT_CREDENTIALS = {
  email: null,
  srpPw: "f6c1cc977d2811c55f0260f0318c8cbe13d215120f5d4f1113b33f32db670e81",
  unwrapBKey: "c8bdcea80dd5ebc94b870f57b840a5f9ea1d82d5eae72a5081831c7f3667be74",
  passwordSalt: "AAAAAA",
  srp: {
    type: "SRP-6a/SHA256/2048/v1",
    salt: "f4f435710b693852e6602a58c902a37539e4ce8351c6495f0b28b4c5cbad0cd4",
    algorithm: "sha256",
    verifier: null,
  }
}


// Build an SRP verifier to go with the dummy credentials defined above.
//
function getDummySRPVerifier(email) {
  var res = srp.getv(
    Buffer(DUMMY_CLIENT_CREDENTIALS.srp.salt, 'hex'),
    Buffer(email),
    Buffer(DUMMY_CLIENT_CREDENTIALS.srpPw),
    srp.params['2048'].N,
    srp.params['2048'].g,
    DUMMY_CLIENT_CREDENTIALS.srp.algorithm
 );
 if (res.length < 256) {
   var tmp = res;
   var padding = res.length - 256;
   var res = new Buffer(256);
   res.full(0, 0, padding);
   tmp.copy(res, padding);
 }
 return res.toString('hex');
}


//  Construct and return a new API Client object.
//  This produces a new Client object with credentials in place, ready
//  to perform API requests and send the resulting performance data back
//  to the loads test runner.  It uses the dummy credentials defined above.
//
//  You must provide the email address to use, and the loads communication
//  socket.
//
function getClient(email, loadsSocket) {

  var client = new Client(SERVER_URL);

  // Set up dummy credentials.
  // We have to mix the email address into the SRP verifier, but
  // everything else can be pre-calculated.
  for (var k in DUMMY_CLIENT_CREDENTIALS) {
    client[k] = DUMMY_CLIENT_CREDENTIALS[k];
  }
  client.email = Buffer(email).toString('hex');
  client.srp.verifier = getDummySRPVerifier(client.email);

  // Instrument the API to send request back back to loads test runner.
  client.api.on('startRequest', function(options) {
    options._loadsStartTime = +new Date();
  });
  client.api.on('endRequest', function(options, err, res) {
    if (!err) {
      var loadsEndTime = +new Date();
      loadsSocket.send('add_hit', {
        url: options.url,
        method: options.method,
        status: res.statusCode,
        started: options._loadsStartTime / 1000,
        elapsed: (loadsEndTime - options._loadsStartTime) / 1000
      });
    }
  });

  return client
}


//  Generate a short random hex string.
//  Useful where you need something unique.
//
function uniqueID(size) {
  if (!size) {
    size = 10;
  }
  return crypto.randomBytes(size/2 + 1).toString('hex').substring(0, size);
}


//  Generate a random integer in the range [min, max]
//
function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}



//  And finally, here is the actual loadtest!
//
function runLoadTest(loadsSocket) {

  // Target ratio is 2 new-account signups per 10 old-account signups.
  // We use a restricted pool of old-account email addresses to simulate
  // existing users; they will be populated quickly during the loadtest
  // run.  We use randomly-generated email addresses to simulate new accounts
  // and will forcibly delete any that are found to exist.

  var email;
  var client;
  var ready;

  if (getRandomInt(1, 12) <= 2) {
    //  New user.
    //  Assume it doesnt exist and try to create; delete and retry if it does.
    email = "loady" + uniqueID() + " @restmail.lcip.org";
    client = getClient(email, loadsSocket);
    ready = client.create().fail(function(err) {
      if (err.code != 400 || err.errno != ERROR_ACCOUNT_EXISTS) {
        throw err;
      }
      return client.destroyAccount().then(function() {
        return client.create();
      });
    }).then(function() {
      return client.login();
    });
  } else {
    //  Existing user.
    //  Assume it exists and try to authentiate; create only if not exists.
    email = "loady" + getRandomInt(1, 100) + " @restmail.lcip.org";
    client = getClient(email, loadsSocket);
    ready = client.login().fail(function(err) {
      if (err.code != 400 || err.errno != ERROR_UNKNOWN_ACCOUNT) {
        throw err;
      }
      return client.create().then(function() {
        return client.login();
      });
    });
  }


  // We fetch the cryptographic key material, then make a bunch of cert
  // signing requests.  The cert-signing request is expected to be by
  // *far* the most frequent operation, so we do a lot of them.

  var done = ready.then(function() {
    // Fetch the key material.
    return client.keys();
  }).then(function(keys) {
    // Check that we actually got some key material.
    assert.equal(typeof keys.kA, 'string', 'kA exists');
    assert.equal(typeof keys.wrapKb, 'string', 'wrapKb exists');
  }).then(function() {
    // Make a bunch of certificate-signing requests.
    // Varying the number of iterations of this loop is a simple way
    // to de-synchronize concurrent runs of this test, spreading out the
    // other operations more evenly throughout the run.
    var p = P();
    for (var i=0; i < getRandomInt(10, 100); i++) {
      p = p.then(function() {
        return client.sign(DUMMY_PUBLIC_KEY, 1000 * 60 * 60);
      }).then(function(cert) {
        assert.equal(typeof cert, 'string', 'cert exists');
      });
    }
    return p;
  });

  return done
}


//  The current loads.js test runner spawns a new process for every run
//  of the test, which can severely hamper throughput.  To work around this
//  for now, we do several iterations of the test before reporting success
//  back to the loads runner.
//
//  XXX TODO: fix this issue upstream in the loads.js runner
//
module.exports.picl_idp_loadtest = function picl_idp_loadtest(cb) {
  var p = P();
  // Try to detect whether we're in a "big" bench suite by looking at
  // the LOADS_STATUS environment variable.  We don't want to loop
  // forever if we're in a quick `make test` run.
  var numLoadRuns = 1;
  if (process.env.LOADS_STATUS && process.env.LOADS_STATUS !== "1,1,1,1") {
    numLoadRuns = 100;
  }
  for (var i=0; i < numLoadRuns; i++) {
    p = p.then(function() {
      return runLoadTest(cb.socket);
    });
  }
  p.done(
    function() { return cb(null); },
    function(err) { return cb(err); }
  );
}

File: packages/fxa-auth-server/loadtest/loadtests.py

import os
import hmac
import json
import math
import random
import hashlib
import urlparse
import hawkauthlib
from base64 import b64encode

from requests.auth import AuthBase

from loads import TestCase


# Error constants used by the fxa-auth-server API.

ERROR_ACCOUNT_EXISTS = 101
ERROR_UNKNOWN_ACCOUNT = 102


# The tests need a public key for the server to sign, but we don't actually
# do anything with it.  It suffices to use a fixed dummy key throughout.

DUMMY_PUBLIC_KEY = {
    'algorithm': 'RS',
    'n': '475938596723561050357149433919674961454460669256778579'
         '095393476820271428065297309134131686299358278907987200'
         '7974809511698859885077002492642203267408776123',
    'e': '65537',
}


def uniq(size=10):
    """Generate a short random hex string."""
    return os.urandom(size // 2 + 1).encode('hex')[:size]


def HKDF_extract(salt, IKM, hashmod=hashlib.sha256):
    """HKDF-Extract; see RFC-5869 for the details."""
    if salt is None:
        salt = b"\x00" * hashmod().digest_size
    return hmac.new(salt, IKM, hashmod).digest()


def HKDF_expand(PRK, info, L, hashmod=hashlib.sha256):
    """HKDF-Expand; see RFC-5869 for the details."""
    digest_size = hashmod().digest_size
    N = int(math.ceil(L * 1.0 / digest_size))
    assert N <= 255
    T = b""
    output = []
    for i in xrange(1, N + 1):
        data = T + info + chr(i)
        T = hmac.new(PRK, data, hashmod).digest()
        output.append(T)
    return b"".join(output)[:L]


def HKDF(secret, salt, info, size, hashmod=hashlib.sha256):
    """HKDF-extract-and-expand as a single function."""
    PRK = HKDF_extract(salt, secret, hashmod)
    return HKDF_expand(PRK, info, size, hashmod)


class HawkAuth(AuthBase):

    def __init__(self, server_url, tokendata, tokentype):
        self.server_url = server_url
        keyInfo = 'identity.mozilla.com/picl/v1/' + tokentype
        keyMaterial = HKDF(tokendata, "", keyInfo, 32*3)
        self.tokenid = keyMaterial[:32]
        self.authKey = keyMaterial[32:64]
        self.bundleKey = keyMaterial[64:]

    def __call__(self, req):
        # Requets doesn't seem to include the port in the Host header,
        # and loads replaces hostnames with IPs.  Undo all this rubbish
        # so that we can calculate the correct signature.
        req.headers['Host'] = urlparse.urlparse(self.server_url).netloc
        id = self.tokenid.encode('hex')
        params = {}
        if req.body:
            payloadStr = 'hawk.1.payload\napplication/json\n'
            payloadStr += req.body + '\n'
            params['hash'] = b64encode(hashlib.sha256(payloadStr).digest())
        hawkauthlib.sign_request(req, id, self.authKey, params=params)
        return req


class LoadTest(TestCase):

    server_url = 'http://api-accounts.loadtest.lcip.org'
    #server_url = 'http://127.0.0.1:9000/'

    def makeurl(self, path):
        return urlparse.urljoin(self.server_url, path)

    def makehawkauth(self, tokendata, tokentype):
        return HawkAuth(self.server_url, tokendata, tokentype)

    def setUp(self):
        super(LoadTest, self).setUp()
        self.credentials = {}
        self.tokens = {}

    def tearDown(self):
        self.tokens = None
        self.credentials = None
        super(LoadTest, self).tearDown()

    def test_auth_server(self):
        self._pick_user_and_authenticate()
        self._fetch_keys()
        self._fetch_random_bytes()
        for i in xrange(random.randint(10, 100)):
            self._sign_public_key()

    def _pick_user_and_authenticate(self):
        email = self._pick_user()
        if "new" in email:
            self._authenticate_as_new_user()
        else:
            self._authenticate_as_existing_user()

    def _pick_user(self):
        # User emails look like this:
        #
        #    loads-<id>-(old|new)@restmail.net
        #
        # The components mean:
        #    * id:   randomly-generated id
        #    * old:  we expect this account to already exist
        #    * new:  we expect this account to not exist
        #
        # We want 2 new-user signups per 10 existing-account logins.
        #
        status = random.choice((['new'] * 2) + (['old'] * 10))
        if status == 'old':
            id = str(random.randint(1, 999))
        else:
            id = uniq()
        email = "loads-%s-%s@restmail.lcip.org" % (id, status)
        self.credentials['email'] = email
        self.credentials['authPW'] = hashlib.sha256(email).hexdigest()
        return email

    def _authenticate_as_new_user(self):
        # Authenticate as a brand-new user account.
        # Assume it doesn't exist, try to create the account.
        # But it's not big deal if it happens to already exist.
        res = self._create_account()
        if res.status_code != 200:
            self.assertEqual(res.status_code, 400)
            err = res.json()
            self.assertEqual(err['errno'], ERROR_ACCOUNT_EXISTS)
        res = self._start_session()
        self.assertEqual(res.status_code, 200)
        return res

    def _authenticate_as_existing_user(self):
        # Authenticate as an existing user account.
        # We select from a small pool of known accounts, creating it
        # if it does not exist.  This should mean that all the accounts
        # are created quickly at the start of the loadtest run.
        res = self._start_session()
        if res.status_code != 200:
            self.assertEqual(res.status_code, 400)
            err = res.json()
            self.assertEqual(err['errno'], ERROR_UNKNOWN_ACCOUNT)
            res = self._create_account()
            # Account creation might fail in turn, due to a race condition.
            if res.status_code != 200:
                self.assertEqual(res.status_code, 400)
                err = res.json()
                self.assertEqual(err['errno'], ERROR_ACCOUNT_EXISTS)
            res = self._start_session()
            self.assertEqual(res.status_code, 200)
        return res

    def _create_account(self):
        res = self._req_POST('/v1/account/create', {
            'email': self.credentials['email'],
            'authPW': self.credentials['authPW'],
            'preVerified': True,
        })
        return res

    def _start_session(self):
        res = self._req_POST('/v1/account/login?keys=true', {
            'email': self.credentials['email'],
            'authPW': self.credentials['authPW'],
        })
        if res.status_code != 200:
            return res
        body = res.json()
        self.assertTrue(body['verified'])
        self.tokens['session'] = body['sessionToken'].decode('hex')
        self.tokens['keyfetch'] = body['keyFetchToken'].decode('hex')
        return res

    def _fetch_keys(self):
        auth = self.makehawkauth(self.tokens['keyfetch'], 'keyFetchToken')
        res = self._req_GET('/v1/account/keys', auth=auth)
        self.assertEqual(res.status_code, 200)
        return res

    def _fetch_random_bytes(self):
        res = self._req_POST('/v1/get_random_bytes', {})
        self.assertEqual(res.status_code, 200)
        self.assertEqual(len(res.json()['data']), 64)
        return res

    def _sign_public_key(self):
        auth = self.makehawkauth(self.tokens['session'], 'sessionToken')
        payload = {
            'publicKey': DUMMY_PUBLIC_KEY,
            'duration': 1000,
        }
        res = self._req_POST('/v1/certificate/sign', payload, auth=auth)
        self.assertEqual(res.status_code, 200)
        return res

    # Raw request-making methods.
    # These are just skinny helpers over the methods on self.session.

    def _req_GET(self, url, auth=None):
        url = self.makeurl(url)
        return self.session.get(url, auth=auth)

    def _req_POST(self, url, payload, auth=None):
        url = self.makeurl(url)
        data = json.dumps(payload)
        headers = {
            'Content-Type': 'application/json',
        }
        return self.session.post(url, data=data, auth=auth, headers=headers)
